test/unit/math/rev/mat/functor/util_algebra_solver.hpp

```diff
@@ -1,12 +1,35 @@
 #include <gtest/gtest.h>
 #include <stan/math/rev/mat/functor/algebra_solver.hpp>
+#include <stan/math/rev/mat/functor/algebra_solver_newton.hpp>
 #include <test/unit/util.hpp>
 #include <sstream>
 #include <vector>
 #include <limits>
 #include <string>
 
-/* define algebraic functions which get solved */
+/* wrapper function that either calls the dogleg or the newton solver. */
+template <typename F, typename T1, typename T2>
+Eigen::Matrix<T2, Eigen::Dynamic, 1> general_algebra_solver(
+    bool is_newton, const F& f, const Eigen::Matrix<T1, Eigen::Dynamic, 1>& x,
```

Comment:
To me it's not a good idea to run different solvers in a single unit test, as this is against the point of "unit test"(I know that I violate this rule sometimes, shame to me).  It's less intuitive to figure out which solver fails if `is_newton` loop reports failure in one of your tests. So it's better to separate than lump: `TEST(xxx, dogleg)` and `TEST(xxx, kinsol)`. To avoid repeating yourself you can take advantage of [test fixture](https://github.com/google/googletest/blob/master/googletest/docs/primer.md#test-fixtures-using-the-same-data-configuration-for-multiple-tests-same-data-multiple-tests).

Another advantage of separate testing + test fixture is performance comparison, since that's one of the reasons why we add Newton.

Reply:
> To me it's not a good idea to run different solvers in a single unit test, as this is against the point of "unit test"

Right. My goal was to avoid adding too much code, but I agree we should prioritize on writing "proper" unit tests. Test Fixture helps, even though there is still a lot of redundant code. See the refactored code.